src/vm/lib/metadata/agent.js

```diff
@@ -228,36 +230,116 @@ function zoneExists(zonename, callback) {
         if (err) {
             if (err.code === 'ENOENT') {
+                opts.log.trace({
+                    sockpath: zoneConn.sockpath
+                }, 'ENOENT on sockpath: stale');
                 callback(null, true); // stale
                 return;
             }
             callback(err);
             return;
         }
-        callback(null, false); // no error in stat means exists: not stale
+
+        // Check for changes in the fs.stat() signature since we created this
+        // socket. If it has changed, that means our handle to it is stale and
+        // we should recreate it.
+        for (field = 0; field < fields.length; field++) {
+            if (zoneConn.sockstat[fields[field]] !== stats[fields[field]]) {
+                opts.log.debug({
+                    field: fields[field],
+                    sockpath: zoneConn.sockpath,
+                    new_sockstat: stats,
+                    old_sockstat: zoneConn.sockstat
+                }, 'change in sockpath fs.stat signature: stale');
+                callback(null, true); // stale
+                return;
+            }
+        }
+        if ((zoneConn.sockstat.ctime
+            && zoneConn.sockstat.ctime.getTime()) !== stats.ctime.getTime()) {
+            opts.log.debug({
+                field: 'ctime',
+                sockpath: zoneConn.sockpath,
+                new_sockstat: stats,
+                old_sockstat: zoneConn.sockstat
+            }, 'change in sockpath fs.stat signature: stale');
+            callback(null, true); // stale
+            return;
+        }
+
+        // no error in stat, and no diff fields means exists: not stale
+        opts.log.trace({
+            sockpath: zoneConn.sockpath
+        }, 'sockpath still exists and fs.stat signature matches: not stale');
+        callback(null, false);
         return;
     });
 }
 
 /*
+ * This function does an fs.stat() on the 'sockpath' of the zoneConn and
+ * attaches the fs.Stats result to the 'zoneConn' object as .sockstat. If there
+ * is an error with fs.stat() conn.sockstat will be set to an empty object.
+ *
+ * After the stat has completed, callback() will be called. Any error from
+ * fs.stat() will be passed as the first and only argument to callback().
+ */
+function addConnSockStat(zoneConn, callback) {
+    assert.object(zoneConn, 'zoneConn');
+    assert.string(zoneConn.sockpath, 'zoneConn.sockpath');
+    assert.func(callback, 'callback');
+
+    fs.stat(zoneConn.sockpath, function _statSock(e, st) {
+        if (e) {
+            // If there was an error w/ the stat, it's most likely because
+            // the state of the world has changed. We'll fill in sockstat
+            // with an empty object here so that checkStaleSocket will report
+            // this as stale.
+            zoneConn.sockstat = {};
+        } else {
+            zoneConn.sockstat = st;
+        }
+        callback(e);
+    });
+}
+
+function closeZoneConnection(zoneConn) {
```

Comment:
I think we should assert that `zoneConn` is an object. I would also suggest asserting that one but only one of `close` or `destroy` methods got called.